# Install Libraries

In [1]:
!pip install --upgrade nbformat

Requirement already up-to-date: nbformat in .\env\lib\site-packages (5.9.2)


You should consider upgrading via the 'd:\an-online-shop-business\env\scripts\python.exe -m pip install --upgrade pip' command.


In [2]:
!pip install scikit-learn
!pip install pandas
!pip install matplotlib
!pip install seaborn
!pip install plotly

You should consider upgrading via the 'd:\an-online-shop-business\env\scripts\python.exe -m pip install --upgrade pip' command.


You should consider upgrading via the 'd:\an-online-shop-business\env\scripts\python.exe -m pip install --upgrade pip' command.


You should consider upgrading via the 'd:\an-online-shop-business\env\scripts\python.exe -m pip install --upgrade pip' command.


You should consider upgrading via the 'd:\an-online-shop-business\env\scripts\python.exe -m pip install --upgrade pip' command.


You should consider upgrading via the 'd:\an-online-shop-business\env\scripts\python.exe -m pip install --upgrade pip' command.


# Importing Libraries

In [3]:
# import sklearn
from sklearn.linear_model import LinearRegression

# Importing for Data Manipulation
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Importing For Data Visualization
import matplotlib.pyplot as plt # for data visualization
import seaborn as sns
# from wordcloud import WordCloud
import plotly.express as px
import plotly.graph_objs as go
from plotly.offline import init_notebook_mode, iplot

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

# import os
# for dirname, _, filenames in os.walk('/kaggle/input'):
#     for filename in filenames:
#         print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All"
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Reading The Dataset

In [4]:
# Reading the dataset with Pandas
df = pd.read_csv("./Sales Transaction v.4a.csv")

# Sampling the data based on heads (5 top ranks)
df.head()

TransactionNo       Date ProductNo                          ProductName  \
0        581482  12/9/2019     22485        Set Of 2 Wooden Market Crates   
1        581475  12/9/2019     22596  Christmas Star Wish List Chalkboard   
2        581475  12/9/2019     23235             Storage Tin Vintage Leaf   
3        581475  12/9/2019     23272    Tree T-Light Holder Willie Winkie   
4        581475  12/9/2019     23239    Set Of 4 Knick Knack Tins Poppies   

   Price  Quantity  CustomerNo         Country  
0  21.47        12     17490.0  United Kingdom  
1  10.65        36     13069.0  United Kingdom  
2  11.53        12     13069.0  United Kingdom  
3  10.65        12     13069.0  United Kingdom  
4  11.94         6     13069.0  United Kingdom

In [5]:
# Sampling the data based on tails (5 least ranks)
df.tail()

TransactionNo       Date ProductNo                   ProductName  \
536345       C536548  12/1/2018     22168  Organiser Wood Antique White   
536346       C536548  12/1/2018     21218        Red Spotty Biscuit Tin   
536347       C536548  12/1/2018     20957  Porcelain Hanging Bell Small   
536348       C536548  12/1/2018     22580  Advent Calendar Gingham Sack   
536349       C536548  12/1/2018     22767    Triple Photo Frame Cornice   

        Price  Quantity  CustomerNo  Country  
536345  18.96        -2     12472.0  Germany  
536346  14.09        -3     12472.0  Germany  
536347  11.74        -1     12472.0  Germany  
536348  16.35        -4     12472.0  Germany  
536349  20.45        -2     12472.0  Germany

# Understanding The Data

#### Based on the source of the data, we can describe the dataset to understanding the meaning of the data. Each data can be describe as follow :
 1. TransactionNo (categorical): a six-digit unique number that defines each transaction. The letter “C” in the code indicates a cancellation.

 2. Date (numeric): the date when each transaction was generated.

 3. ProductNo (categorical): a five or six-digit unique character used to identify a specific product.

 4. Product (categorical): product/item name.

 5. Price (numeric): the price of each product per unit in pound sterling (£).

 6. Quantity (numeric): the quantity of each product per transaction. Negative values related to cancelled transactions.

 7. CustomerNo (categorical): a five-digit unique number that defines each customer.

 8. Country (categorical): name of the country where the customer resides.


In [6]:
# Print the shape of the dataset
df.shape

(536350, 8)

In [7]:
# Print to describing about the dataset
df.describe()

Price       Quantity     CustomerNo
count  536350.000000  536350.000000  536295.000000
mean       12.662182       9.919347   15227.893178
std         8.490450     216.662300    1716.582932
min         5.130000  -80995.000000   12004.000000
25%        10.990000       1.000000   13807.000000
50%        11.940000       3.000000   15152.000000
75%        14.090000      10.000000   16729.000000
max       660.620000   80995.000000   18287.000000

In [8]:
# Print to look for info about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [9]:
# Print to searching about missing values
print ("Missing Values")
print ("-"*30)
print(df.isnull().sum())

Missing Values
------------------------------
TransactionNo     0
Date              0
ProductNo         0
ProductName       0
Price             0
Quantity          0
CustomerNo       55
Country           0
dtype: int64


In [10]:
# Searching for data duplicates
print ("Data Duplicatioin")
print ("-"*30)
print(df.duplicated(keep=False).sum())

Data Duplicatioin
------------------------------
9994


In [11]:
# Check the unique value on Dataset
print("Unique Values")
print("-"*30)
print(df.nunique())

Unique Values
------------------------------
TransactionNo    23204
Date               305
ProductNo         3768
ProductName       3768
Price              530
Quantity           510
CustomerNo        4738
Country             38
dtype: int64


In [12]:
print(f"There's {df['Country'].nunique()} Country in this Dataset")
print("-"*30)
print('Including :', df['Country'].unique())

There's 38 Country in this Dataset
------------------------------
Including : ['United Kingdom' 'Norway' 'Belgium' 'Germany' 'France' 'Austria'
 'Netherlands' 'EIRE' 'USA' 'Channel Islands' 'Iceland' 'Portugal' 'Spain'
 'Finland' 'Italy' 'Greece' 'Japan' 'Sweden' 'Denmark' 'Cyprus' 'Malta'
 'Switzerland' 'Australia' 'Czech Republic' 'Poland' 'Hong Kong'
 'Singapore' 'RSA' 'Israel' 'Unspecified' 'United Arab Emirates' 'Canada'
 'European Community' 'Bahrain' 'Brazil' 'Saudi Arabia' 'Lebanon'
 'Lithuania']


From here we learn that :

1. The dataset contain (536350 rows of transaction, 8 collumns of attributes).
2. The 'Date' collumn stored as an object (text), which is not a proper date and time format
3. The 'CustomerNo' collumn stored as an float numbers, which can cause ambiguousity
4. Total of country that this database have is 38 Country. Which mean this online shop business have transaction across the country in the world.
5. Some data have missing values, including the CustomerNo, which is unexpected. Every transaction mustly have a CustomerNo in there.
6. In Quantity collumn, there have negative values. It can cause ambigousity to.
7. In Transaction collumn, we must remember that if theres have letter "C" in the first value. That indicate cancellation.

# Data Cleaning

In [13]:
# Display the rows in the 'CostumerNo' collumn:
rows_in_costumerno_missing = df[df['CustomerNo'].isnull()]
print ("Rows CostumerNo With Missing Values")
rows_in_costumerno_missing.head()

Rows CostumerNo With Missing Values


TransactionNo        Date ProductNo  \
6511         C581406   12/8/2019    46000M   
6512         C581406   12/8/2019    46000S   
90098        C575153   11/8/2019     22947   
102671       C574288   11/3/2019     22178   
117263       C573180  10/28/2019     23048   

                                ProductName  Price  Quantity  CustomerNo  \
6511           Polyester Filler Pad 45x45cm   6.19      -240         NaN   
6512           Polyester Filler Pad 40x40cm   6.19      -300         NaN   
90098            Wooden Advent Calendar Red  44.25        -1         NaN   
102671      Victorian Glass Hanging T-Light  25.37        -1         NaN   
117263  Set Of 10 Lanterns Fairy Light Star  14.50        -1         NaN   

               Country  
6511    United Kingdom  
6512    United Kingdom  
90098   United Kingdom  
102671  United Kingdom  
117263  United Kingdom

In [14]:
#cleaning the data with missing values
df = df.dropna()
#checking the data
print ("Missing Values")
print ("-"*30)
print(df.isnull().sum())

Missing Values
------------------------------
TransactionNo    0
Date             0
ProductNo        0
ProductName      0
Price            0
Quantity         0
CustomerNo       0
Country          0
dtype: int64


In [15]:
#check if there's duplicates in dataset
print ('Data Duplicates')
print ("-"*30)
print (df.duplicated(keep=False).sum())

Data Duplicates
------------------------------
9994


In [16]:
#cleaning the duplicates
df = df.drop_duplicates()

In [17]:
#checking again if there's duplicates in dataset
print ('Data Duplicates')
print ("-"*30)
print (df.duplicated(keep=False).sum())

Data Duplicates
------------------------------
0


In [18]:
#check description of the databased
df.describe()

Price       Quantity     CustomerNo
count  531095.000000  531095.000000  531095.000000
mean       12.669635       9.993146   15222.612241
std         8.526181     217.710261    1716.633588
min         5.130000  -80995.000000   12004.000000
25%        10.990000       1.000000   13798.000000
50%        11.940000       3.000000   15146.000000
75%        14.090000      10.000000   16727.000000
max       660.620000   80995.000000   18287.000000

> if we look at the database, there have negative value in 'Quantity'. We must clean that
and make sure that TransactionNo must'nt have 'C' indicator in there (Cancellation).

In [19]:
# Searching the TransactionNo with 'C' indicators
cancelation = df[df['TransactionNo'].str.startswith('C') & (df['Quantity'] < 0)]

# Print the cancelation rows
print ("Cancelation Orders")
cancelation

Cancelation Orders


TransactionNo       Date ProductNo                       ProductName  \
1616         C581484  12/9/2019     23843         Paper Craft Little Birdie   
1617         C581490  12/9/2019     22178   Victorian Glass Hanging T-Light   
1618         C581490  12/9/2019     23144   Zinc T-Light Holder Stars Small   
1619         C581568  12/9/2019     21258        Victorian Sewing Box Large   
1620         C581569  12/9/2019     84978  Hanging Heart Jar T-Light Holder   
...              ...        ...       ...                               ...   
536345       C536548  12/1/2018     22168      Organiser Wood Antique White   
536346       C536548  12/1/2018     21218            Red Spotty Biscuit Tin   
536347       C536548  12/1/2018     20957      Porcelain Hanging Bell Small   
536348       C536548  12/1/2018     22580      Advent Calendar Gingham Sack   
536349       C536548  12/1/2018     22767        Triple Photo Frame Cornice   

        Price  Quantity  CustomerNo         Country  
1616     6.19    -80995     16446.0  United Kingdom  
1617     6.19       -12     14397.0  United Kingdom  
1618     6.04       -11     14397.0  United Kingdom  
1619     6.19        -5     15311.0  United Kingdom  
1620     6.19        -1     17315.0  United Kingdom  
...       ...       ...         ...             ...  
536345  18.96        -2     12472.0         Germany  
536346  14.09        -3     12472.0         Germany  
536347  11.74        -1     12472.0         Germany  
536348  16.35        -4     12472.0         Germany  
536349  20.45        -2     12472.0         Germany  

[8494 rows x 8 columns]

let's go delete them because there are invalid values

In [20]:
# Delete Cancelation Transaction based on Quantity below 0
df = df[df['Quantity'] > 0]

In [21]:
# Checking the databased again
df.describe()

Price       Quantity     CustomerNo
count  522601.000000  522601.000000  522601.000000
mean       12.637160      10.667492   15226.311767
std         7.965974     157.542420    1716.555479
min         5.130000       1.000000   12004.000000
25%        10.990000       1.000000   13804.000000
50%        11.940000       4.000000   15152.000000
75%        14.090000      12.000000   16729.000000
max       660.620000   80995.000000   18287.000000

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 522601 entries, 0 to 536324
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  522601 non-null  object 
 1   Date           522601 non-null  object 
 2   ProductNo      522601 non-null  object 
 3   ProductName    522601 non-null  object 
 4   Price          522601 non-null  float64
 5   Quantity       522601 non-null  int64  
 6   CustomerNo     522601 non-null  float64
 7   Country        522601 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 35.9+ MB


In [23]:
# Changing the type 'Date' to datetime
df['Date'] = pd.to_datetime(df['Date'])

In [24]:
# Changing the data type 'CustomerNo' and 'TransactionId' into integer
df['CustomerNo'] = df['CustomerNo'].astype(int)
df['TransactionNo'] = df['TransactionNo'].astype(int)

In [25]:
#Let's check it again
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 522601 entries, 0 to 536324
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   TransactionNo  522601 non-null  int32         
 1   Date           522601 non-null  datetime64[ns]
 2   ProductNo      522601 non-null  object        
 3   ProductName    522601 non-null  object        
 4   Price          522601 non-null  float64       
 5   Quantity       522601 non-null  int64         
 6   CustomerNo     522601 non-null  int32         
 7   Country        522601 non-null  object        
dtypes: datetime64[ns](1), float64(1), int32(2), int64(1), object(3)
memory usage: 31.9+ MB


# Exploratory Data Analysis (EDA)

First we need to set the time of database for being analyze

In [26]:
# Setting the date and time
start_date = df.Date.min()
end_date = df.Date.max()
period = end_date - start_date

print("Start:", start_date)
print("End:", end_date)
print("Period:", period)

Start: 2018-12-01 00:00:00
End: 2019-12-09 00:00:00
Period: 373 days 00:00:00


In [27]:
# Get orders counts of countries
countries = df['Country'].value_counts()

# Prepare data dictionary for choropleth map
data = dict(type='choropleth', locations=countries.index, locationmode='country names', z=countries, text=countries.index, colorbar={'title': 'Order num.'})

# Define layout for choropleth map
layout = dict(title='Total number of orders per country', geo=dict(showframe=True, projection={'type': 'mercator'}), width=1000, height=1000)

# Create choropleth map figure
choromap = go.Figure(data=[data], layout=layout)

# Display the choropleth map
iplot(choromap, validate=False)

In [28]:
# Creating Revenue Collumn
df['Revenue'] = df['Price'] * df['Quantity']
df.sample()

TransactionNo       Date ProductNo                  ProductName  Price  \
15357         580697 2019-12-05     23058  Christmas Hanging Snowflake   5.97   

       Quantity  CustomerNo         Country  Revenue  
15357         2       15984  United Kingdom    11.94

In [29]:
# Grouping Country and Revenue
top_revenue = df.groupby(['Country'])['Revenue'].sum().reset_index()
top_revenue = top_revenue.sort_values(by=['Revenue'], ascending = False)

# Creating Diagram Bar to determine which country has top revenue
fig = px.bar(top_revenue.head(5), x='Country', y='Revenue', color='Revenue', title='Highest Revenue Countries')
fig.update_xaxes(title='Country')
fig.update_yaxes(title='Revenue')
fig.show()

as you guys can see... the highest revenue in this databased is United Kingdom with **52346795**.

In [30]:
# Searching Total Product in Database
print('Total Product on this Database:\n' + str(df['ProductName'].nunique()))

Total Product on this Database:
3753


In [31]:
# Grouping ProductName with Quantity
top_product = df.groupby(['ProductName'])['Quantity'].sum().reset_index()
top_product = top_product.sort_values(by=['Quantity'], ascending = False)

# Creating Diagram Bar to determine which country has top revenue
fig = px.bar(top_product.head(5), x='ProductName', y='Quantity',title='Top Product Based on Quantity')
fig.update_xaxes(title='Product Name')
fig.update_yaxes(title='Total Quantity Sold')
fig.show()

In [32]:
# Grouping ProductName with Revenue
top_product = df.groupby(['ProductName'])['Revenue'].sum().reset_index()
top_product = top_product.sort_values(by=['Revenue'], ascending = False)

# Creating Diagram Bar to determine which country has top revenue
fig = px.bar(top_product.head(10), x='ProductName', y='Revenue', color='Revenue', title='Top Product Based on Revenue')
fig.update_xaxes(title='Product Name')
fig.update_yaxes(title='Total Revenue')
fig.show()

In [33]:
# Searching Total Customer in Database
print('Total Customer on this Database:\n' + str(df['CustomerNo'].nunique()))

Total Customer on this Database:
4718


In [34]:
# Grouping CustomerNo with Quantity
top_customer_by_quantity = df.groupby(['CustomerNo'])['Quantity'].sum().reset_index()
top_customer_by_quantity = top_customer_by_quantity.sort_values(by=['Quantity'], ascending = False)

# Horizontal bar plot for top customers by quantity
fig_customer_quantity = px.bar(top_customer_by_quantity.head(10), x="Quantity", y="CustomerNo", color="Quantity", orientation='h', title="Top Customers according to Product Quantity Sold")
fig_customer_quantity.update_xaxes(title="Quantity")
fig_customer_quantity.update_yaxes(title="CustomerNo", type = 'category')
fig_customer_quantity.show()

In [35]:
# Grouping CustomerNo with Revenue
top_customer_by_revenue = df.groupby(['CustomerNo'])['Revenue'].sum().reset_index()
top_customer_by_revenue = top_customer_by_revenue.sort_values(by=['Revenue'], ascending = False)

# Horizontal bar plot for top customers by quantity
fig_customer_revenue = px.bar(top_customer_by_revenue.head(10), x="Revenue", y="CustomerNo", color="Revenue", orientation='h', title="Top Customers according to Revenue")
fig_customer_revenue.update_xaxes(title="Revenue")
fig_customer_revenue.update_yaxes(title="CustomerNo", type = 'category')
fig_customer_revenue.show()